File: samples/msgext-action-quickstart/python/bots/bot_activity_handler.py
from botbuilder.core import ActivityHandler, TurnContext, CardFactory
from botbuilder.schema import HeroCard, InvokeResponse
from botbuilder.schema._models_py3 import CardImage


class BotActivityHandler(ActivityHandler):
    async def on_invoke_activity(self, turn_context: TurnContext):
        """
        Handle invoke activities (e.g., messaging extension actions).
        """
        activity = turn_context.activity
        if activity.name == "composeExtension/submitAction":
            action = activity.value
            if action.get("commandId") == "createCard":
                return await self.create_card_command(turn_context, action)
            elif action.get("commandId") == "shareMessage":
                return await self.share_message_command(turn_context, action)
            else:
                return self._create_invoke_response(
                    status=400, body={"error": "Command not implemented"}
                )
        elif activity.name == "composeExtension/queryLink":
            return await self.handle_query_link(turn_context, activity.value)
        return await super().on_invoke_activity(turn_context)
    
    async def handle_query_link(self, turn_context: TurnContext, query):
        """
        Handle unfurling links for messaging extensions.
        """
        attachment = CardFactory.thumbnail_card(
            title="Thumbnail Card",
            images=[query["url"]],
            buttons=[
                {
                    "type": "openUrl",
                    "title": "View Package",
                    "value": "https://www.npmjs.com/package/level-option-wrap",
                }
            ],
        )
        result = {
            "composeExtension": {
                "type": "result",
                "attachmentLayout": "list",
                "attachments": [attachment],
            }
        }
        return self._create_invoke_response(status=200, body=result)


    async def create_card_command(self, context: TurnContext, action: dict) -> InvokeResponse:
        """
        Create a card in response to the 'createCard' action.
        """
        data = action.get("data", {})
        hero_card = HeroCard(
            title=data.get("title", ""),
            text=data.get("text", ""),
            subtitle=data.get("subTitle", "")
        )

        # Create an attachment using CardFactory
        attachment = CardFactory.hero_card(hero_card)

        # Serialize the attachment for JSON serialization
        serialized_attachment = attachment.serialize()

        result = {
            "composeExtension": {
                "type": "result",
                "attachmentLayout": "list",
                "attachments": [serialized_attachment],
            }
        }
        return self._create_invoke_response(status=200, body=result)

    async def share_message_command(self, context: TurnContext, action: dict) -> InvokeResponse:
        """
        Share a message in response to the 'shareMessage' action.
        """
        user_name = "unknown"
        message_payload = action.get("messagePayload", {})
        if "from" in message_payload and "user" in message_payload["from"]:
            user_name = message_payload["from"]["user"].get("displayName", user_name)

        images = []
        include_image = action.get("data", {}).get("includeImage", "false")
        if include_image == "true":
            images = [CardImage(url="https://encrypted-tbn0.gstatic.com/images?q=tbn:ANd9GcQtB3AwMUeNoq4gUBGe6Ocj8kyh3bXa9ZbV7u1fVKQoyKFHdkqU")]

        # Create the HeroCard
        hero_card = HeroCard(
            title=f"{user_name} originally sent this message:",
            text=message_payload.get("body", {}).get("content", ""),
            subtitle=f"({len(message_payload.get('attachments', []))} attachments not included)"
            if "attachments" in message_payload
            else None,
            images=images,
        )

        # Convert HeroCard to an attachment using CardFactory
        attachment = CardFactory.hero_card(hero_card)
        serialized_attachment = attachment.serialize()

        result = {
            "composeExtension": {
                "type": "result",
                "attachmentLayout": "list",
                "attachments": [serialized_attachment],
            }
        }

        # Return InvokeResponse with the result
        return InvokeResponse(status=200, body=result)
    
    def _create_invoke_response(self, status: int, body: dict) -> InvokeResponse:
        """
        Helper to create an InvokeResponse.
        """
        return InvokeResponse(status=status, body=body)

File: samples/msgext-action-quickstart/js/botActivityHandler.js
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

const {
    TurnContext,
    MessageFactory,
    TeamsActivityHandler,
    CardFactory,
    ActionTypes
} = require('botbuilder');

class BotActivityHandler extends TeamsActivityHandler {
    constructor() {
        super();
    }
    /* Building a messaging extension action command is a two-step process.
        (1) Define how the messaging extension will look and be invoked in the client.
            This can be done from the Configuration tab, or the Manifest Editor.
            Learn more: https://aka.ms/teams-me-design-action.
        (2) Define how the bot service will respond to incoming action commands.
            Learn more: https://aka.ms/teams-me-respond-action.
        
        NOTE:   Ensure the bot endpoint that services incoming messaging extension queries is
                registered with Bot Framework.
                Learn more: https://aka.ms/teams-register-bot.
    */

    // Invoked when the service receives an incoming action command.
    handleTeamsMessagingExtensionSubmitAction(context, action) {
        /* Commands are defined in the manifest file. This can be done using the Configuration tab, editing
           the manifest Json directly, or using the manifest editor in App Studio. This project includes two
           commands to help get you started: createCard and shareMessage.
        */
        switch (action.commandId) {
        case 'createCard':
            return createCardCommand(context, action);
        case 'shareMessage':
            return shareMessageCommand(context, action);
        default:
            throw new Error('NotImplemented');
        }
    }

    /* Messaging Extension - Unfurling Link */
    handleTeamsAppBasedLinkQuery(context, query) {
        const attachment = CardFactory.thumbnailCard('Thumbnail Card',
            query.url,
            ['https://raw.githubusercontent.com/microsoft/botframework-sdk/master/icon.png']);
    
        const result = {
            attachmentLayout: 'list',
            type: 'result',
            attachments: [attachment]
        };
    
        const response = {
            composeExtension: result
        };
        return response;
        }
    /* Messaging Extension - Unfurling Link */
}

function createCardCommand(context, action) {
    // The user has chosen to create a card by choosing the 'Create Card' context menu command.
    const data = action.data;
    const heroCard = CardFactory.heroCard(data.title, data.text);
    heroCard.content.subtitle = data.subTitle;
    const attachment = { contentType: heroCard.contentType, content: heroCard.content, preview: heroCard };

    return {
        composeExtension: {
            type: 'result',
            attachmentLayout: 'list',
            attachments: [
                attachment
            ]
        }
    };
}

function shareMessageCommand(context, action) {
    // The user has chosen to share a message by choosing the 'Share Message' context menu command.
    let userName = 'unknown';
    if (action.messagePayload.from &&
            action.messagePayload.from.user &&
            action.messagePayload.from.user.displayName) {
        userName = action.messagePayload.from.user.displayName;
    }

    // This Messaging Extension example allows the user to check a box to include an image with the
    // shared message.  This demonstrates sending custom parameters along with the message payload.
    let images = [];
    const includeImage = action.data.includeImage;
    if (includeImage === 'true') {
        images = ['https://encrypted-tbn0.gstatic.com/images?q=tbn:ANd9GcQtB3AwMUeNoq4gUBGe6Ocj8kyh3bXa9ZbV7u1fVKQoyKFHdkqU'];
    }
    const heroCard = CardFactory.heroCard(`${ userName } originally sent this message:`,
        action.messagePayload.body.content,
        images);

    if (action.messagePayload.attachments && action.messagePayload.attachments.length > 0) {
        // This sample does not add the MessagePayload Attachments.  This is left as an
        // exercise for the user.
        heroCard.content.subtitle = `(${ action.messagePayload.attachments.length } Attachments not included)`;
    }

    const attachment = { contentType: heroCard.contentType, content: heroCard.content, preview: heroCard };

    return {
        composeExtension: {
            type: 'result',
            attachmentLayout: 'list',
            attachments: [
                attachment
            ]
        }
    };
}

module.exports.BotActivityHandler = BotActivityHandler;

